applications/GeoMechanicsApplication/tests/cpp_tests/test_time_incrementor.cpp

```diff
@@ -0,0 +1,439 @@
+    const auto start_increment   = 0.5;
+    const auto max_num_of_cycles = std::size_t{4};
+    const auto time_incrementor  = AdaptiveTimeIncrementor{start_time, end_time, start_increment, max_num_of_cycles};
+    auto previous_state          = TimeStepEndState{};
+    previous_state.convergence_state = TimeStepEndState::ConvergenceState::non_converged;
+
+    KRATOS_EXPECT_FALSE(time_incrementor.WantRetryStep(max_num_of_cycles, previous_state))
+}
+
+KRATOS_TEST_CASE_IN_SUITE(DontRetryWhenPreviousAttemptConverged, KratosGeoMechanicsFastSuite)
+{
+    const auto start_time        = 0.0;
+    const auto end_time          = start_time + 8.0;
+    const auto start_increment   = 0.5;
+    const auto max_num_of_cycles = std::size_t{4};
+    const auto time_incrementor  = AdaptiveTimeIncrementor{start_time, end_time, start_increment, max_num_of_cycles};
+    const auto cycle_number      = std::size_t{1};
+    auto previous_state          = TimeStepEndState{};
+    previous_state.convergence_state = TimeStepEndState::ConvergenceState::converged;
+
+    KRATOS_EXPECT_FALSE(time_incrementor.WantRetryStep(cycle_number, previous_state))
+}
+
+KRATOS_TEST_CASE_IN_SUITE(GetStartIncrementWhenItWouldNotResultInExceedingTheEndTime, KratosGeoMechanicsFastSuite)
+{
+    const auto start_time        = 0.0;
+    const auto end_time          = start_time + 8.0;
+    const auto start_increment   = 0.5;
+    const auto max_num_of_cycles = std::size_t{4};
+    const auto reduction_factor  = 0.5;
+    const auto time_incrementor  = AdaptiveTimeIncrementor{start_time, end_time, start_increment, max_num_of_cycles, reduction_factor};
+
+    KRATOS_EXPECT_DOUBLE_EQ(start_increment, time_incrementor.GetIncrement());
+}
+
+KRATOS_TEST_CASE_IN_SUITE(ReduceStartIncrementWhenItWouldResultInExceedingTheEndTime, KratosGeoMechanicsFastSuite)
+{
+    const auto start_time       = 0.0;
+    const auto start_increment  = 1.0;
+    const auto end_time         = start_time + 0.5 * start_increment; // `end_time` would be exceeded if `start_increment` is applied
+    const auto time_incrementor = AdaptiveTimeIncrementor{start_time, end_time, start_increment};
+
+    KRATOS_EXPECT_DOUBLE_EQ(end_time - start_time, time_incrementor.GetIncrement());
+}
+
+KRATOS_TEST_CASE_IN_SUITE(ReduceIncrementWhenPreviousAttemptDidNotConverge, KratosGeoMechanicsFastSuite)
+{
+    const auto start_time        = 0.0;
+    const auto end_time          = start_time + 8.0;
+    const auto start_increment   = 0.5;
+    const auto max_num_of_cycles = std::size_t{4};
+    const auto reduction_factor  = 0.6;
+    auto time_incrementor        = AdaptiveTimeIncrementor{start_time, end_time, start_increment, max_num_of_cycles, reduction_factor};
+    auto previous_state          = TimeStepEndState{};
+    previous_state.convergence_state = TimeStepEndState::ConvergenceState::non_converged;
+
+    time_incrementor.PostTimeStepExecution(previous_state); // previous attempt did not converge
```

Comment:
unexpected comment. PostTimeStepExecution processes the end state and sets up for the new cycle or increment.

Reply:
Done.